core/psh/psh.c

```diff
@@ -89,15 +91,25 @@ int main(int argc, char **argv)
 	while (write(1, "", 0) < 0)
 		usleep(50000);
 
-	base = basename(argv[0]);
+	/* Check if its first shell */
+	pid = tcgetpgrp(STDIN_FILENO);
```

Comment:
not a bad idea, but it breaks using psh as a preinit (which is done in DTR project) - in syspage we have `psh -i /etc/rc.psh` which never ends now (so the device is not booting).

Let's remove not-exiting-from-top-psh from this PR, as it should be taken care of in the same PR as `isrunpsh` hack removal. Regardless, I would stick with explicit commandline param (or explicit `argv[0]`) to specify non-exitable psh.

Reply:
I do not know DTR project internals but if it uses `psh -i` then it is not interactive `psh` so it does not take terminal control (`tcgetpgrp` output remains constant). We can later solve this by using different loop condition:
```
while (tcpid == -1 && tcgetpgrp(STDIN_FILENO) == getpgrp());
```
but I'll follow your suggestion and use nonexitability exclusively for `pshlogin` with:
```
 while (tcpid == -1 && strcmp(base, "pshlogin") == 0);
```

The login forcing using parameter (e.g `psh -l`) will be considered in `ispshrun` hack PR